### DISCUSSION TITLE
Ability to Extend Action Buttons

### DISCUSSION THREAD

**Discussion by eliteproxy7** [directus/directus#4707] [Feature Requests]
I want to add some custom action buttons on the collection layout page near the icons on top ... something like this

![image](https://user-images.githubusercontent.com/5486268/112734236-e511c200-8f1a-11eb-8323-7e2ac21d8b5c.png)

and when I click on that button it performs an action on all the selected items. 

Or perhaps opens the side panel with additional options to be used to achieve the bulk interactions with selected products I am looking for.

is this something that is possible currently? 

Do I have to either modify the core code? i really don't want to create a custom extension layout that duplicates everything the default one has and do my modifications on that

**Comment by philleepflorence** [comment]
Hello @eliteproxy7 ... Start with the Concept Documentation for [Extensions](https://docs.directus.io/concepts/extensions/) and [Layouts](https://docs.directus.io/concepts/layouts/).

The [Directus Tabular Layout ](https://github.com/directus/directus/tree/main/app/src/layouts/tabular) can be a Guide.

**Reply by t7tran** [reply]
We was considering this approach. However, it seems that we will unknowningly end up copying Directus source code - which is something we would like to avoid.

Another concern is that `useLayoutState` isn't available in the custom layout.

**Comment by rijkvanzanten** [comment]
There's currently no way to extend just the action buttons 👍🏻

**Reply by benhaynes** [reply]
Correct. Though this topic has come up in the past. I'm not 100% sure how I feel about opening this area up, since it's a critical area for system actions... though I'm open to discussing it. That being said, the Info Sidebar _is_ intended to be extensible, and those are also available via 1-click (and on any page). So perhaps for now we push forward with formalizing that extension type?

**Reply by t7tran** [reply]
We're also interested in this feature. It streamlines the additonal actions to be done on the item, i.e. Export to PDF, Route User to a related custom module page, etc. So far, we can only see using Presentation Links field as a workaround which seems to always open a new browser tab.

**Comment by jpetryk** [comment]
My use case definitely fits extending the action buttons rather than using the sidebar, simply because my use case is definitely an extra "action". Any thoughts on not allowing the removal or changing of any of the existing actions because they are standard Directus actions, but providing a way to insert one or more actions per collection? So not necessarily on the layout screen as was originally described here, but simply a configuration associated with the Collection Setup that requires one to define the various button states and then simply fires a webhook (assumes the trigger action is Update or something). This would allow users to explicitly trigger webhooks that call other systems, rather than relying on internal events.

More details on my specific use case: we are fine with that current actions do and don't want to fire a webhook on the current "Save" action. We'd like that to only have the effect of saving inside Directus as it already does. But upon clicking a new button, available at all times, it will trigger a "Build".

Perhaps best described as **a manual webhook tied to a custom action** that can be associated with a collection.

**Reply by benhaynes** [reply]
I think only supporting an external webhook from custom action buttons might be a bit too limiting. But I like that as an option. I think these should be custom extensions (or a slot) that allow for anything (added to what's there already). So I guess the extension would define a route that the button should exist on, and the action logic that should happen: custom code, webhook, etc.

**Reply by eliteproxy7** [reply]
So we assign the "Action Button" from the collection's settings. which exposes its action as a trigger source in Webhooks?

Currently, the way i achieved my feature was to create a dropdown field and call a webhook based on that field change and performed desired actions through the API.

so the dropdown field contains a set of actions and based on an update to the field my external API performed whatever actions I needed, including talking about to directus via the API to update someting. 

Maybe it would be good to make the Webhooks part of the broader set called Actions and this way you can later expand to custom code, navigation, integrations

**Comment by that1matt** [comment]
We are also interested in this functionality being extensible. Our use case, is that we integrate with an external service to translate content on all of collections, where we can send a single item, or a batch job with multiple collection items. 

Having this functionality to at least hook into, to easily get all collection item ids that were selected, to be sent off to our webhook to get the collection item fields, their content values, to send off to our translation service would be extremely helpful.

**Reply by rijkvanzanten** [reply]
Directus Flows has a manual trigger 🙂 The only difference is that the button shows up in the right sidebar instead of header actions, but otherwise it's meant for this use case

**Reply by that1matt** [reply]
The downside to this, is that we have a lot of collections, in the manual trigger flow option, there is no way to select all, or filter based on a set of criteria where a collection should or should not be included in the `collections` list. Having to manually track, enable/disable is more cumbersome, compared to showing/not showing an action on supported/unsupported collections.

**Reply by JonL1** [reply]
A 3 dotted button that opens a dropdown and lists the flows for that collection with the option of filtering them would be cool :)

![Safari 08-05-2023 10 32 08 000457](https://user-images.githubusercontent.com/1325712/236776312-500095c3-582f-4cfb-8cec-cd6095300562.png)

**Reply by rijkvanzanten** [reply]
> The downside to this, is that we have a lot of collections, in the manual trigger flow option, there is no way to select all, or filter based on a set of criteria where a collection should or should not be included in the `collections` list. Having to manually track, enable/disable is more cumbersome, compared to showing/not showing an action on supported/unsupported collections.

I think we should reenvision how that config is stored and used. Instead of an array of exact matches, we could store it as glob pattern, that way you could use a `*` for everything, or `prefix-*` for everything with a given prefix. We'll have to figure out what the right UX for that is though, as it might be a bit too technical for some, and having to type out 3-of-5 collections is a bit of a chore the other way 🤔

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
